olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/provenance/io/ProvenanceMarshaller.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates.
+ *
+ * Licensed under the 2-clause BSD license.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.oracle.labs.mlrg.olcut.provenance.io;
+
+import com.oracle.labs.mlrg.olcut.provenance.ObjectProvenance;
+import com.oracle.labs.mlrg.olcut.provenance.ProvenanceUtil;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+
+/**
+ * Interface for serializing and deserializing marshalled provenances to
+ * and from strings or files.
+ */
+public interface ProvenanceMarshaller {
```

Comment:
I'm not sure if this is the right name for this interface based on the Javadoc description.  I see that it can do double-duty for both marshalling and serializing, but the marshalling part is for convenience. Mostly this interface exists to allow implementors to stitch in the serialization part, rather than the marshalling to the intermediate form. Maybe ProvenanceSerializer instead? I'd also consider ProvenanceSerialization which (in my mind) says it provides a serialization ability, which in turn implies that it both serializes and deserializes. Not that Serializer precludes that, but it might make somebody look for a ProvenanceDeserializer too if they weren't clear about what this does. Just some food for thought here though.

Reply:
I think I'm fine with `ProvenanceSerialization`. Following through that would rename the others to `JsonProvenanceSerialization` and `XMLProvenanceSerialization` which is a little odd, so `ProvenanceSerializer` might work better?

Reply:
Oh there's already `JsonProvenanceSerializer` because Jackson's interface is `StdSerializer`. Not sure what's a better name.

Reply:
I don't think `JsonProvenanceSerialization` and `XMLProvenanceSerialization` are too awkward, but having it right next to JsonProvenanceSerializer does seem pretty confusing. Is ProvenanceWriter any better? It doesn't exactly work with the Java nomenclature for Writer though. ProvenanceFormat or ProvenanceFormatter?